README.md
# cwlogs-rqid-filter.py
_Ever wanted to filter AWS CloudWatch logs and not only keep the matching events, but also all events that have the same Request ID that the matching event(s)?_

This python script fetches all log events messages related to requests (by AWS Request ID) that in any message of any event match a custom python regex pattern.
It fetches all events for the period, searching their messages with the custom regex pattern and filters only events that have the request IDs that have at least one message matching.

Changes are described in CHANGELOG.md.

## Installation
`pip install cwlogs-rqid-filter`

## Usage
Do not forget to perform AWS Credentials configuration for boto3 beforehand (https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html).

### Console
Install the library as described above, and run it, either with `cwlogs-rqid-filter` or `python -m cwlogs_rqid_filter`


```
    python -m cwlogs_rqid_filter [-h] --group GROUP_NAME --filter FILTER
                                 [--streams [STREAM_NAMES [STREAM_NAMES ...]]]
                                 [--stream-prefix STREAM_NAME_PREFIX]
                                 [--start-ts START_TIMESTAMP] [--start START]
                                 [--stop-ts STOP_TIMESTAMP] [--stop STOP] [--limit LIMIT]
                                 [--prefix-timestamp | --prefix-iso] [--debug]


Filter AWS CloudWatch logs while keeping all events that have the same Request
ID as events that match the filter.

optional arguments:
  -h, --help            show this help message and exit
  --group GROUP_NAME    Log group name
  --filter FILTER       Python regular expression pattern that will match
                        events messages
  --streams [STREAM_NAMES [STREAM_NAMES ...]]
                        Log stream names
  --stream-prefix STREAM_NAME_PREFIX
                        Log stream name prefix
  --start-ts START_TIMESTAMP
                        Start timestamp in milliseconds, UTC timezone
  --start START         Start date and time, ISO8601 format, UTC timezone
  --stop-ts STOP_TIMESTAMP
                        Stop timestamp in milliseconds, UTC timezone
  --stop STOP           Stop date and time, ISO8601 format, UTC timezone
  --limit LIMIT         Event limit count
  --prefix-timestamp    Prefix the logs with event timestamp between
                        parentheses
  --prefix-iso          Prefix the logs with ISO8601 formatted event timestamp
                        between parentheses
  --debug               Print debug information
```

### Script import
```
from cwlogs_rqid_filter import fetch_events, filter_events

# Any parameters set accepted by the Boto3 logs client filter_log_events() function
request_parameters = {
    'logGroupName': 'xxx',
    'logStreamNames': ['yyy', 'zzz', ...],
    'logStreamNamePrefix': 'xyz',
    'startTime': 123,                       # Unix timestamp
    'endTime': 456,                         # Unix timestamp
    ...
}

filter_regex_pattern = r'*'

all_events = fetch_events(request_parameters)
filtered_events = filter_events(all_events, filter_regex_pattern)
```

## Examples
To get all log events of lambda requests that took at least 1000ms, prefixed by ISO8601-formatted timestamps:
`python -m cwlogs_rqid_filter --group /aws/lambda/someLambdaName --start 2018-11-30T05:04:00Z --stop 2018-11-30T05:05:00Z --filter 'Billed Duration: [0-9]{4,}' --prefix-iso`

You can also specify start and stop timestamps in any timezone, formatted following ISO8601: `--start 2018-11-30T14:04:00+09:00 --stop 2018-11-30T14:05:00+09:00`


CHANGELOG.md
# Changelog
All notable thanges to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
None

## [0.2.1] - 2019-07-17
### Changed
- Update description

## [0.2.0] - 2019-07-17
### Added
- Added `cwlogs-rqid-filter` script
- Added this changelog file

### Fixed
- Fixed typos in package name


cwlogs_rqid_filter/console.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
from datetime import datetime

import ciso8601

from . import fetch_events, filter_events, logger


def _add_arguments(parser):
    parser.add_argument(
        "--group", dest="group_name", type=str, required=True, help="Log group name"
    )
    parser.add_argument(
        "--filter",
        dest="filter",
        type=str,
        required=True,
        help="Python regular expression pattern that will match events messages",
    )
    parser.add_argument(
        "--streams",
        dest="stream_names",
        type=str,
        nargs="*",
        required=False,
        help="Log stream names",
    )
    parser.add_argument(
        "--stream-prefix",
        dest="stream_name_prefix",
        type=str,
        required=False,
        help="Log stream name prefix",
    )
    parser.add_argument(
        "--start-ts",
        dest="start_timestamp",
        type=int,
        required=False,
        help="Start timestamp in milliseconds, UTC timezone",
    )
    parser.add_argument(
        "--start",
        dest="start",
        type=str,
        required=False,
        help="Start date and time, ISO8601 format, UTC timezone",
    )
    parser.add_argument(
        "--stop-ts",
        dest="stop_timestamp",
        type=int,
        required=False,
        help="Stop timestamp in milliseconds, UTC timezone",
    )
    parser.add_argument(
        "--stop",
        dest="stop",
        type=str,
        required=False,
        help="Stop date and time, ISO8601 format, UTC timezone",
    )
    parser.add_argument(
        "--limit", dest="limit", type=int, required=False, help="Event limit count"
    )
    timestamp_prefix_group = parser.add_mutually_exclusive_group(required=False)
    timestamp_prefix_group.add_argument(
        "--prefix-timestamp",
        dest="prefix_ts",
        action="store_true",
        help="Prefix the logs with event timestamp between parentheses",
    )
    timestamp_prefix_group.add_argument(
        "--prefix-iso",
        dest="prefix_iso",
        action="store_true",
        help="Prefix the logs with ISO8601 formatted event timestamp between parentheses",
    )
    parser.add_argument(
        "--debug", dest="debug", action="store_true", help="Print debug information"
    )


def run():
    # Arguments handling
    import argparse

    description = (
        "Filter AWS CloudWatch logs while keeping all events that have the same Request ID as "
        "events that match the filter."
    )
    usage = """
    python -m cwlogs_rqid_filter [-h] --group GROUP_NAME --filter FILTER
                                 [--streams [STREAM_NAMES [STREAM_NAMES ...]]]
                                 [--stream-prefix STREAM_NAME_PREFIX]
                                 [--start-ts START_TIMESTAMP] [--start START]
                                 [--stop-ts STOP_TIMESTAMP] [--stop STOP] [--limit LIMIT]
                                 [--prefix-timestamp | --prefix-iso] [--debug]
    """

    parser = argparse.ArgumentParser(description=description, usage=usage)

    _add_arguments(parser)
    args = parser.parse_args()

    if args.debug:
        logger.setLevel("DEBUG")

    # Build request parameters
    request_parameters = {"logGroupName": args.group_name}
    if args.stream_names:
        request_parameters["logStreamNames"] = args.stream_names
    if args.stream_name_prefix:
        request_parameters["logStreamNamePrefix"] = args.stream_name_prefix
    if args.limit:
        request_parameters["limit"] = args.limit

    if args.start_timestamp:
        request_parameters["startTime"] = args.start_timestamp
    elif args.start:
        request_parameters["startTime"] = int(
            ciso8601.parse_datetime(args.start).timestamp() * 1000
        )

    if args.stop_timestamp:
        request_parameters["endTime"] = args.stop_timestamp
    elif args.stop:
        request_parameters["endTime"] = int(
            ciso8601.parse_datetime(args.stop).timestamp() * 1000
        )

    # Run
    events = fetch_events(request_parameters)
    events = filter_events(events, args.filter)

    # Print events prefixed with their timestamp
    print_pattern = "({}) {}"

    for event in events:
        timestamp = event["timestamp"]

        if args.prefix_ts:
            print(print_pattern.format(timestamp, event["message"]))
        elif args.prefix_iso:
            iso8601_ts = datetime.utcfromtimestamp(timestamp // 1000).replace(
                microsecond=timestamp % 1000 * 1000
            )
            print(print_pattern.format(iso8601_ts, event["message"]))
        else:
            print(event["message"])


cwlogs_rqid_filter/__init__.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import logging
import re

import boto3

logging.basicConfig()
logger = logging.getLogger(__name__)
logger.setLevel("INFO")

client = boto3.client("logs")

# This pattern matches AWS Request IDs of format xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
# with 'x's any lowercase 0-9, a-f character
REQUEST_ID_REGEX = re.compile(
    r"(([a-f]|\d){8}-([a-f]|\d){4}-([a-f]|\d){4}-([a-f]|\d){4}-([a-f]|\d){12})"
)


def fetch_events(request_parameters):
    """
    Fetches events from AWS CloudWatch and returns a list of timestamp-sorted events.
    Basically a wrapper around boto3 logs filter_log_events() that takes care of collecting all
    data behind every nextToken.

    Args:
        request_parameters (dict): supported Boto3 logs client filter_log_events() arguments
            {
                'logGroupName': 'xxx',
                'logStreamNames': ['yyy', 'zzz', ...],
                'logStreamNamePrefix': 'xyz',
                'startTime': 123,
                'endTime': 456,
                ...
            }

    Returns:
        list: list of event dicts
    """
    events = []
    next_token = None

    while True:
        if next_token:
            request_parameters["nextToken"] = next_token

        logger.debug("Performing query with parameters: {}".format(request_parameters))
        response = client.filter_log_events(**request_parameters)

        response_events = response["events"]
        logger.debug("Got {} events from this response".format(len(response_events)))

        events += response_events
        next_token = response.get("nextToken")

        if not next_token:
            searched_log_streams = response["searchedLogStreams"]
            searched_log_stream_names = [
                s["logStreamName"] for s in searched_log_streams
            ]
            completely_searched_log_stream_names = [
                s["logStreamName"]
                for s in searched_log_streams
                if s["searchedCompletely"]
            ]

            break

    # Sort events by timestamp
    events = sorted(events, key=lambda x: x["timestamp"])

    logger.debug("Retrieved {} events".format(len(events)))
    logger.debug("Searched log streams {}".format(searched_log_stream_names))
    logger.debug(
        "Completely searched log streams {}".format(
            completely_searched_log_stream_names
        )
    )

    return events


def filter_events(events, filter_regex_pattern):
    """
    Filters events using the provided python regex pattern, and returns not only events that match,
    but also all other events of the same request ID.

    Args:
        events (list): list of events dicts
        filter_regex_pattern (str): string python regex pattern

    Returns:
        list: filtered list of event dicts
    """
    logger.debug("Using filter pattern {}".format(filter_regex_pattern))

    # Build filter regex
    filter_regex = re.compile(filter_regex_pattern)

    # Collect request IDs that match the pattern and write request IDs into events
    matching_request_ids = set()
    for event in events:
        # Collect, strip and replace message
        message = event["message"].strip()
        event["message"] = message

        # Get request ID
        request_id = REQUEST_ID_REGEX.search(message)
        if request_id:
            request_id = request_id.group(0)
        else:
            continue

        event["request_id"] = request_id

        # Match filter pattern
        match = True if filter_regex.search(message) else False

        if match:
            matching_request_ids.add(request_id)

    # Collect events that have a matching request ID
    filtered_events = []
    for event in events:
        if "request_id" in event and event["request_id"] in matching_request_ids:
            filtered_events.append(event)

    logger.debug("Filtered {} events".format(len(filtered_events)))

    return filtered_events


cwlogs_rqid_filter/__main__.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
from datetime import datetime

import ciso8601

from . import fetch_events, filter_events, logger


def _add_arguments(parser):
    parser.add_argument(
        "--group", dest="group_name", type=str, required=True, help="Log group name"
    )
    parser.add_argument(
        "--filter",
        dest="filter",
        type=str,
        required=True,
        help="Python regular expression pattern that will match events messages",
    )
    parser.add_argument(
        "--streams",
        dest="stream_names",
        type=str,
        nargs="*",
        required=False,
        help="Log stream names",
    )
    parser.add_argument(
        "--stream-prefix",
        dest="stream_name_prefix",
        type=str,
        required=False,
        help="Log stream name prefix",
    )
    parser.add_argument(
        "--start-ts",
        dest="start_timestamp",
        type=int,
        required=False,
        help="Start timestamp in milliseconds, UTC timezone",
    )
    parser.add_argument(
        "--start",
        dest="start",
        type=str,
        required=False,
        help="Start date and time, ISO8601 format, UTC timezone",
    )
    parser.add_argument(
        "--stop-ts",
        dest="stop_timestamp",
        type=int,
        required=False,
        help="Stop timestamp in milliseconds, UTC timezone",
    )
    parser.add_argument(
        "--stop",
        dest="stop",
        type=str,
        required=False,
        help="Stop date and time, ISO8601 format, UTC timezone",
    )
    parser.add_argument(
        "--limit", dest="limit", type=int, required=False, help="Event limit count"
    )
    timestamp_prefix_group = parser.add_mutually_exclusive_group(required=False)
    timestamp_prefix_group.add_argument(
        "--prefix-timestamp",
        dest="prefix_ts",
        action="store_true",
        help="Prefix the logs with event timestamp between parentheses",
    )
    timestamp_prefix_group.add_argument(
        "--prefix-iso",
        dest="prefix_iso",
        action="store_true",
        help="Prefix the logs with ISO8601 formatted event timestamp between parentheses",
    )
    parser.add_argument(
        "--debug", dest="debug", action="store_true", help="Print debug information"
    )


if __name__ == "__main__":
    # Arguments handling
    import argparse

    description = (
        "Filter AWS CloudWatch logs while keeping all events that have the same Request ID as "
        "events that match the filter."
    )
    usage = """
    python -m cwlogs-rqid-filter [-h] --group GROUP_NAME --filter FILTER
                                 [--streams [STREAM_NAMES [STREAM_NAMES ...]]]
                                 [--stream-prefix STREAM_NAME_PREFIX]
                                 [--start-ts START_TIMESTAMP] [--start START]
                                 [--stop-ts STOP_TIMESTAMP] [--stop STOP] [--limit LIMIT]
                                 [--prefix-timestamp | --prefix-iso] [--debug]
    """

    parser = argparse.ArgumentParser(description=description, usage=usage)

    _add_arguments(parser)
    args = parser.parse_args()

    if args.debug:
        logger.setLevel("DEBUG")

    # Build request parameters
    request_parameters = {"logGroupName": args.group_name}
    if args.stream_names:
        request_parameters["logStreamNames"] = args.stream_names
    if args.stream_name_prefix:
        request_parameters["logStreamNamePrefix"] = args.stream_name_prefix
    if args.limit:
        request_parameters["limit"] = args.limit

    if args.start_timestamp:
        request_parameters["startTime"] = args.start_timestamp
    elif args.start:
        request_parameters["startTime"] = int(
            ciso8601.parse_datetime(args.start).timestamp() * 1000
        )

    if args.stop_timestamp:
        request_parameters["endTime"] = args.stop_timestamp
    elif args.stop:
        request_parameters["endTime"] = int(
            ciso8601.parse_datetime(args.stop).timestamp() * 1000
        )

    # Run
    events = fetch_events(request_parameters)
    events = filter_events(events, args.filter)

    # Print events prefixed with their timestamp
    print_pattern = "({}) {}"

    for event in events:
        timestamp = event["timestamp"]

        if args.prefix_ts:
            print(print_pattern.format(timestamp, event["message"]))
        elif args.prefix_iso:
            iso8601_ts = datetime.utcfromtimestamp(timestamp // 1000).replace(
                microsecond=timestamp % 1000 * 1000
            )
            print(print_pattern.format(iso8601_ts, event["message"]))
        else:
            print(event["message"])


cwlogs-rqid-filter.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Retrieve and filter AWS CloudWatch logs while keeping all events that contain the same Request ID as
events that match the filter.
"""

import logging
import re
from datetime import datetime

import boto3
import ciso8601

logging.basicConfig()
logger = logging.getLogger(__name__)
logger.setLevel("INFO")

client = boto3.client("logs")

# This pattern matches AWS Request IDs of format xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
# with 'x's any lowercase 0-9, a-f character
REQUEST_ID_REGEX = re.compile(
    r"(([a-f]|\d){8}-([a-f]|\d){4}-([a-f]|\d){4}-([a-f]|\d){4}-([a-f]|\d){12})"
)


def fetch_events(request_parameters):
    """
    Fetches events from AWS CloudWatch and returns a list of timestamp-sorted events.
    Basically a wrapper around boto3 logs filter_log_events() that takes care of collecting all
    data behind every nextToken.

    Args:
        request_parameters (dict): supported filter_log_events() arguments
            {
                'logGroupName': 'xxx',
                'logStreamNames': ['yyy', 'zzz', ...],
                'logStreamNamePrefix': 'xyz',
                'startTime': 123,
                'endTime': 456,
                'limit': 678
            }

    Returns:
        list: list of event dicts
    """
    events = []
    next_token = None

    while True:
        if next_token:
            request_parameters["nextToken"] = next_token

        logger.debug("Performing query with parameters: {}".format(request_parameters))
        response = client.filter_log_events(**request_parameters)

        response_events = response["events"]
        logger.debug("Got {} events from this response".format(len(response_events)))

        events += response_events
        next_token = response.get("nextToken")

        if not next_token:
            searched_log_streams = response["searchedLogStreams"]
            searched_log_stream_names = [
                s["logStreamName"] for s in searched_log_streams
            ]
            completely_searched_log_stream_names = [
                s["logStreamName"]
                for s in searched_log_streams
                if s["searchedCompletely"]
            ]

            break

    # Sort events by timestamp
    events = sorted(events, key=lambda x: x["timestamp"])

    logger.debug("Retrieved {} events".format(len(events)))
    logger.debug("Searched log streams {}".format(searched_log_stream_names))
    logger.debug(
        "Completely searched log streams {}".format(
            completely_searched_log_stream_names
        )
    )

    return events


def filter_events(events, filter_regex_pattern):
    """
    Filters events using the provided python regex pattern, and returns not only events that match,
    but also all other events of the same request ID.

    Args:
        events (list): list of events dicts
        filter_regex_pattern (str): string python regex pattern

    Returns:
        list: filtered list of event dicts
    """
    logger.debug("Using filter pattern {}".format(filter_regex_pattern))

    # Build filter regex
    filter_regex = re.compile(filter_regex_pattern)

    # Collect request IDs that match the pattern and write request IDs into events
    matching_request_ids = set()
    for event in events:
        # Collect, strip and replace message
        message = event["message"].strip()
        event["message"] = message

        # Get request ID
        request_id = REQUEST_ID_REGEX.search(message)
        if request_id:
            request_id = request_id.group(0)
        else:
            continue

        event["request_id"] = request_id

        # Match filter pattern
        match = True if filter_regex.search(message) else False

        if match:
            matching_request_ids.add(request_id)

    # Collect events that have a matching request ID
    filtered_events = []
    for event in events:
        if "request_id" in event and event["request_id"] in matching_request_ids:
            filtered_events.append(event)

    logger.debug("Filtered {} events".format(len(filtered_events)))

    return filtered_events


def _add_arguments(parser):
    parser.add_argument(
        "--group", dest="group_name", type=str, required=True, help="Log group name"
    )
    parser.add_argument(
        "--filter",
        dest="filter",
        type=str,
        required=True,
        help="Python regular expression pattern that will match events messages",
    )
    parser.add_argument(
        "--streams",
        dest="stream_names",
        type=str,
        nargs="*",
        required=False,
        help="Log stream names",
    )
    parser.add_argument(
        "--stream-prefix",
        dest="stream_name_prefix",
        type=str,
        required=False,
        help="Log stream name prefix",
    )
    parser.add_argument(
        "--start-ts",
        dest="start_timestamp",
        type=int,
        required=False,
        help="Start timestamp in milliseconds, UTC timezone",
    )
    parser.add_argument(
        "--start",
        dest="start",
        type=str,
        required=False,
        help="Start date and time, ISO8601 format, UTC timezone",
    )
    parser.add_argument(
        "--stop-ts",
        dest="stop_timestamp",
        type=int,
        required=False,
        help="Stop timestamp in milliseconds, UTC timezone",
    )
    parser.add_argument(
        "--stop",
        dest="stop",
        type=str,
        required=False,
        help="Stop date and time, ISO8601 format, UTC timezone",
    )
    parser.add_argument(
        "--limit", dest="limit", type=int, required=False, help="Event limit count"
    )
    timestamp_prefix_group = parser.add_mutually_exclusive_group(required=False)
    timestamp_prefix_group.add_argument(
        "--prefix-timestamp",
        dest="prefix_ts",
        action="store_true",
        help="Prefix the logs with event timestamp between parentheses",
    )
    timestamp_prefix_group.add_argument(
        "--prefix-iso",
        dest="prefix_iso",
        action="store_true",
        help="Prefix the logs with ISO8601 formatted event timestamp between parentheses",
    )
    parser.add_argument(
        "--debug", dest="debug", action="store_true", help="Print debug information"
    )


if __name__ == "__main__":
    # Arguments handling
    import argparse

    description = (
        "Filter AWS CloudWatch logs while keeping all events that have the same Request ID as "
        "events that match the filter."
    )
    parser = argparse.ArgumentParser(description=description)
    _add_arguments(parser)
    args = parser.parse_args()

    if args.debug:
        logger.setLevel("DEBUG")

    # Build request parameters
    request_parameters = {"logGroupName": args.group_name}
    if args.stream_names:
        request_parameters["logStreamNames"] = args.stream_names
    if args.stream_name_prefix:
        request_parameters["logStreamNamePrefix"] = args.stream_name_prefix
    if args.limit:
        request_parameters["limit"] = args.limit

    if args.start_timestamp:
        request_parameters["startTime"] = args.start_timestamp
    elif args.start:
        request_parameters["startTime"] = int(
            ciso8601.parse_datetime(args.start).timestamp() * 1000
        )

    if args.stop_timestamp:
        request_parameters["endTime"] = args.stop_timestamp
    elif args.stop:
        request_parameters["endTime"] = int(
            ciso8601.parse_datetime(args.stop).timestamp() * 1000
        )

    # Run
    events = fetch_events(request_parameters)
    events = filter_events(events, args.filter)

    # Print events prefixed with their timestamp
    print_pattern = "({}) {}"

    for event in events:
        timestamp = event["timestamp"]

        if args.prefix_ts:
            print(print_pattern.format(timestamp, event["message"]))
        elif args.prefix_iso:
            iso8601_ts = datetime.utcfromtimestamp(timestamp // 1000).replace(
                microsecond=timestamp % 1000 * 1000
            )
            print(print_pattern.format(iso8601_ts, event["message"]))
        else:
            print(event["message"])
